app/models/manageiq/providers/ibm_cloud/power_virtual_servers/cloud_manager/template.rb

```diff
@@ -47,6 +49,87 @@ def raw_delete_image
     _log.error("image=[#{name}], error: #{e}")
   end
 
+  def self.encrypt_with_aes(creds)
+    cipher = OpenSSL::Cipher.new('aes-256-cbc')
+    cipher.encrypt
+
+    key  = Base64.strict_encode64(cipher.random_key)
+    iv   = Base64.strict_encode64(cipher.random_iv)
+    encr = Base64.strict_encode64(cipher.update(creds.to_json) + cipher.final)
+
+    [encr, key, iv]
+  end
+
+  def self.miq_img_by_ids(provider_id, image_id)
+    powervc = ExtManagementSystem.find(provider_id)
+    powervc.get_image_info(image_id)
+  end
+
+  def self.cos_creds(provider_id)
+    cos = ExtManagementSystem.find(provider_id)
+    cos.cos_creds
+  end
+
+  def self.node_creds(provider_id)
+    powervc = ExtManagementSystem.find(provider_id)
+    endp = powervc.node_endpoint
+    auth = powervc.node_auth
+
+    return endp.hostname, auth.userid, auth.password
+  end
+
+  def self.image_ems_ref(bucket_id)
+    MiqTemplate.find(bucket_id).uid_ems
+  end
+
+  def self.bucket_name(bucket_id)
+    CloudObjectStoreContainer.find(bucket_id).name
+  end
+
+  def self.raw_import_image(ext_management_system, options = {})
+    session_id = DateTime.now.strftime('%Q')
+
+    location, user, password = node_creds(options['src_provider_id'])
+    hosts = ["[powervc]\n#{location}\n[powervc:vars]\nansible_connection=ssh\nansible_user=#{user}\nansible_ssh_pass=#{password}"]
+
+    cos_credentials = cos_creds(options['obj_storage_id'])
+    bucket = bucket_name(options['bucket_id'])
+
+    diskType = CloudVolumeType.find(options['disk_type_id']).name
+
+    cos_ans_creds = {:resource_instance_id => cos_credentials[0], :apikey => cos_credentials[1], :bucket_name => bucket, :url_endpoint => cos_credentials[3]}
+    cos_pvs_creds = {:region => cos_credentials[2], :bucketName => bucket, :accessKey => cos_credentials[4], :secretKey => cos_credentials[5]}
+
+    encr_cos_creds, encr_cos_key, encr_cos_iv = encrypt_with_aes(cos_ans_creds)
```

Comment:
Ansible runner has handling of credential files already, not sure that encrypting the string then passing the decryption key in the same extra_vars is dramatically more secure :shrug:

Reply:
The ansible playbook turns the creds into a file on the powervc node while the "key" and "iv" are turned into environmental variables for the python process we use on the powervc side. This won't let a simple user decrypt the credentials since the user doesn't have access to the environment of the python process. You can see it here: content/ansible_runner/roles/image-copy-pvc-to-pvs/defaults/main.yml. Not sure I can achieve the sane functionality using the built-in "credentials" option.

Reply:
See also the following three lines: https://github.com/iv1111/manageiq-providers-ibm_cloud/blob/739f5f36f2aad71364dbe8229442399fba3f2434/content/ansible_runner/roles/image-copy-pvc-to-pvs/files/upload_ova_to_bucket.py#L16